Better codestyle enforcement

DennisSimon opened the pull request:
Hi @PhilippWendler ,

this is a quality of life improvement as well as a better way of enforcing code style in the project. 

I have added a couple of things: 
- eslint and prettier configuration files
- `lint` script to check codestyle on all files in scope. This also now gets run in the gitlab linting step
- `lint:fix` script to add a way to let eslint fix all files in scope, this also runs prettier on all stylesheets and json files
- `husky` hook to automatically fix staged `.js`, `.json`, `.css` and `.scss` files and check for linting errors
- create `dependency.json` in gitlab to prevent eslint from throwing because of missing file.

The  commits b000fb0 and c83b539 include the results of a run of `lint:fix` as well as manual fixes of all remaining warnings and errors.

Looking forward to your feedback :)

PhilippWendler commented on .gitlab-ci.yml:
```diff
@@ -43,37 +43,37 @@ stages:
   <<: *unit-tests
   variables:
-    PYTHON_VERSION: '2.7'
+    PYTHON_VERSION: "2.7"
```
Where do these formatting changes to the Yaml file come from? Is there now a linter for that? (Where?)

DennisSimon replied:
I think they automatically got applied when saving the file because the prettier plugin of code kicked in.

PhilippWendler commented on .gitlab-ci.yml:
```diff
@@ -145,15 +143,15 @@ javascript-eslint:
   before_script:
     - cd benchexec/tablegenerator/react-table
     - npm install
+    - npm run dependencies
```
This should be run automatically if it is necessary, just like we do it for the other build jobs.

DennisSimon replied:
I added a specific script to the package.json that does that

PhilippWendler commented on benchexec/tablegenerator/react-table/.eslintrc:
```diff
@@ -0,0 +1,16 @@
+{
```
There is already an `eslintConfig` in `package.json`. We should not have two places where this is configured, so I suggest to move this to `package.json`.

DennisSimon replied:
Will do

PhilippWendler commented on benchexec/tablegenerator/react-table/.prettierrc:
```diff
@@ -0,0 +1,5 @@
+{
```
The point of having an opinionated code formatter is to have no need for thinking about and discussing its configuration. In fact, a good opinionated code formatter would not even provide options at all (e.g., google-java-format). So I would prefer to keep the default config unless there are really strong arguments.

A line length of 160 is way too much anyway, line lengths should be kept to 80-100 characters. But the default line length is better than others.

PhilippWendler commented on benchexec/tablegenerator/react-table/package.json:
```diff
@@ -19,9 +19,41 @@
+    "lint:fix:css": "prettier --write $(pwd)'/**/*.css'",
+    "lint:fix:scss": "prettier --write $(pwd)'/**/*.scss'",
+    "lint:fix:json": "prettier --write $(pwd)'/**/*.json'",
+    "lint:fix": "npm run lint:fix:js && npm run lint:fix:css && npm run lint:fix:scss && npm run lint:fix:json",
```
I guess we could simplify this into a single rule? I see no need for ever calling prettier only on a subset of files.

(Same for rule `lint` as well, of course.)

DennisSimon replied:
Totally agree, cleaned it up.

PhilippWendler commented on benchexec/tablegenerator/react-table/package.json:
```diff
@@ -19,9 +19,41 @@
+      "pre-commit": "lint-staged"
+    }
+  },
+  "lint-staged": {
```
This partially duplicates the above rules, why? Also, it seems inconsistent, e.g., it runs `git add` only for `*.js`, and it uses `eslint-friendly-formatter` that is not referenced anywhere else.

DennisSimon replied:
I also cleaned this up a bit.
Note that this `lint-staged` part is not a script but a configuration for the `lint-staged` package.
`eslint-friendly-formatter` is a terminal friendly wrapper for eslint. 
When you try to verify code with eslint in the console, you will see that it spams the terminal. This tool gets rid of that and makes it easier to find mistakes.

PhilippWendler commented on benchexec/tablegenerator/react-table/package.json:
```diff
@@ -123,12 +157,14 @@
     "file-loader": "^4.2.0",
     "fs-extra": "^8.1.0",
     "html-webpack-plugin": "^4.0.0-beta.8",
+    "husky": "^4.2.3",
```
Isn't `husky` more a development tool that one would install locally on your machine (like an editor) instead of per project?

DennisSimon replied:
Not really as husky is a tool for git hooks and therefore must/should be present whenever we run git commands. 
The configuration for said hooks are defined in the ´package.json´ and might be husky version specific. 
It is therefore recommended to install husky as a dev dependency to enforce its presence and version.

PhilippWendler replied:
Hm, I am not yet convinced about the value of this. In the current config, it only runs a subset of our checks, so developers still need to remember to run other checks manually. But running more checks automatically would soon start to annoy developers if it takes too much time everytime they want to create a temporary commit etc. Furthermore, the JS project is only a part of the repo, and running JS checks on commits that do not touch the JS code is useless.

So I suggest we drop this and developers who like git hooks can of course just configure them locally on their own without much disadvantage.

DennisSimon replied:
In the current configuration it would run `lint:fix` and then would add the linted files to the commit stage again. 
This would happen automatically for the extensions specified in the `lint:fix` script (namely css, scss, json, js). 
It would also be possible to configure the hook in way where it would only run on files in the `react-tables` subfolder. 
This was mainly an idea to make sure that the ci doesn't fail because of a missing trailing comma or a missing space after open brackets. 
I do however strongly agree on the fact that the JavaScript part of this project is not the main focus and therefore may have implications on people not working with this (sub)module. 
I will remove the hook, and clean up the whole branch.

DennisSimon commented:
I addressed the review comments and rebased to master. 
I think the built files changed because i changed some variable names from snake_case to camelCase.

DennisSimon commented on .gitlab-ci.yml:
```diff
@@ -146,8 +146,7 @@ javascript-eslint:
     - cd benchexec/tablegenerator/react-table
     - npm install
   script:
-    - npx eslint src --max-warnings 0
```
Accidentally removed this as well. 
Do you think we should move this to the script?
Will change after I finish work today

DennisSimon commented on benchexec/tablegenerator/react-table/package.json:
```diff
@@ -23,8 +23,23 @@
     "build": "node scripts/dependencies.js && node scripts/build.js",
-    "test": "(test -f src/data/dependencies.json || node scripts/dependencies.js) && node scripts/test.js"
+    "test": "(test -f src/data/dependencies.json || node scripts/dependencies.js) && node scripts/test.js",
+    "lint:fix": "prettier-eslint --write $(pwd)'/**/*.js' && prettier --write $(pwd)'/**/*.+(js|css|scss|json)'",
```
Forgot to remove `js` from prettier

PhilippWendler commented on benchexec/tablegenerator/react-table/package.json:
```diff
@@ -33,7 +48,25 @@
-      "prettier/prettier": "warn"
+      "prettier/prettier": "warn",
+      "no-console": "off",
+      "no-underscore-dangle": "off",
```
Why are some rules turned off here? We currently have no warnings, so it should not be necessary to disable any?

PhilippWendler commented on benchexec/tablegenerator/react-table/src/App.js:
```diff
@@ -10,11 +10,13 @@ import Overview from "./components/Overview";
 
 if (process.env.NODE_ENV !== "production") {
   // load example data for development
+  // eslint-disable-next-line global-require, import/no-unresolved
```
Why do we now need more disabling of warnings?

PhilippWendler commented on benchexec/tablegenerator/react-table/src/components/LinkOverlay.js:
```diff
@@ -28,11 +28,10 @@ export default class LinkOverlay extends React.Component {
   //    (this fails for ZIPs on the local disk).
   // 3) Try loading url from within ZIP archive without Range header.
   loadContent = async url => {
-    console.log("load content", url);
```
The logging of this method has been quite useful in various circumstances, please do not remove it.

PhilippWendler commented on benchexec/tablegenerator/react-table/src/components/QuantilePlot.js:
```diff
@@ -205,7 +212,7 @@ export default class QuantilePlot extends React.Component {
           const data = this[task + i];
           const id = getRunSetName(tool);
-          this.lineCount++;
+          this.lineCount += 1;
```
What is the purpose of changes like this?

DennisSimon replied:
That change seems to have happened because of this eslint rule:
eslint.org/docs/rules/no-plusplus

PhilippWendler commented on benchexec/tablegenerator/react-table/src/components/QuantilePlot.js:
```diff
@@ -291,11 +301,11 @@ export default class QuantilePlot extends React.Component {
     );
     const type = this.state.isValue ? this.possibleValues[index].type : null;
 
+    const interpolation = this.state.linear ? "linear" : "log";
```
This does not have anything to do with interpolation.

DennisSimon commented:
Moved to #556.
